netbox/dcim/models/devices.py

```diff
@@ -958,10 +958,10 @@ def clean(self):
                 })
 
         # A Device can only be assigned to a Cluster in the same Site (or no Site)
-        if self.cluster and self.cluster.site is not None and self.cluster.site != self.site:
+        if self.cluster and self.cluster._site is not None and self.cluster._site != self.site:
```

Comment:
This should be extended to validate against the cluster's assigned region/site group/location as well.